### PR TITLE
feat(MixedMultiPicker): Allow secondaryOptions to be cleared and reset on a primary option

### PR DESCRIPTION
**Pull request by EthanMcM**
## **Description**

Allow secondaryOptions to be cleared and reset on a primary option

#### **Verify that...**

- [X ] Any related demos were added and `npm start` and `npm run build` still works
- [X ] New demos work in `Safari`, `Chrome` and `Firefox`
- [X ] `npm run lint` passes
- [X ] `npm test` passes and code coverage is increased
- [ X] `npm run build` still works

#### **Bullhorn Internal Developers**
- [x] Run `Novo Automation`

##### **Screenshots**

### CODE REVIEW

File: projects/novo-elements/src/elements/picker/extras/mixed-multi-picker-results/MixedMultiPickerResults.ts [hunk 1]
```diff
@@ -5,7 +5,7 @@ import { BasePickerResults } from '../base-picker-results/BasePickerResults';
 import { Helpers } from '../../../../utils/Helpers';
 import { NovoLabelService } from '../../../../services/novo-label-service';
 import { NovoListElement } from '../../../list/List';
-import { fromEvent, Subscription } from 'rxjs';
+import { BehaviorSubject, fromEvent, Observable, Subscription } from 'rxjs';
```
Comment (trenthibbard) [inline projects/novo-elements/src/elements/picker/extras/mixed-multi-picker-results/MixedMultiPickerResults.ts:8]: `Observable` here is an unused import

File: projects/novo-elements/src/elements/picker/extras/mixed-multi-picker-results/MixedMultiPickerResults.ts [hunk 2]
```diff
@@ -166,6 +167,13 @@ export class MixedMultiPickerResults extends BasePickerResults implements OnDest
         return !!(primaryOption && primaryOption.showSearchOnSecondaryOptions);
     }
 
+    public resetPrimaryOption(primaryOption: IMixedMultiPickerOption) {
+        if (this.internalMap.get(primaryOption.value)) {
+            this.internalMap.delete(primaryOption.value);
+            this.selectPrimaryOption(primaryOption);
```
Comment (trenthibbard) [inline projects/novo-elements/src/elements/picker/extras/mixed-multi-picker-results/MixedMultiPickerResults.ts:173]: Do we always want this option to be selected whenever the BehaviorSubject emits an event? I could see use cases in the future where we want to clear the cached secondary options of several primary options, and in that case we wouldn't want to select them all after

**Review by trenthibbard — CHANGES_REQUESTED** [review]

**Review by trenthibbard — APPROVED** [review]

**Review by trenthibbard — APPROVED** [review]

File: projects/novo-elements/src/elements/picker/extras/mixed-multi-picker-results/MixedMultiPickerResults.ts [hunk 3]
```diff
@@ -17,6 +17,7 @@ export interface IMixedMultiPickerOption {
         filterValue?: any,
     }[];
     getSecondaryOptionsAsync?(): Promise<{value: string, label: string}[]>;
+    clearSecondaryOptions?: BehaviorSubject<Boolean>;
```
Comment (amrutha-rajiv) [inline projects/novo-elements/src/elements/picker/extras/mixed-multi-picker-results/MixedMultiPickerResults.ts:20]: Why is `clearSecondaryOptions` a BehaviorSubject? I'm curious about how this is used by the component, it might also help to update the examples in novo-elements for the MuxesMultiPickerResults component.

File: projects/novo-elements/src/elements/picker/extras/mixed-multi-picker-results/MixedMultiPickerResults.ts [hunk 4]
```diff
@@ -166,6 +167,18 @@ export class MixedMultiPickerResults extends BasePickerResults implements OnDest
...
+            if (primaryOption.value === this.selectedPrimaryOption?.value) {
+                this.activeMatch = null;
+                this.matches = [];
+                this.selectedPrimaryOption = null;
+            }
+            this.internalMap.delete(primaryOption.value);
+            this.ref.markForCheck();
+        }
+    }
+
```
[the remark this answers is not in the document]
Comment (EthanMcM) [reply projects/novo-elements/src/elements/picker/extras/mixed-multi-picker-results/MixedMultiPickerResults.ts:181]: unit tests were added for this!

File: projects/novo-elements/src/elements/form/FieldInteractionApi.ts [hunk 5]
```diff
@@ -591,7 +591,7 @@ export class FieldInteractionApi {
         if (index !== -1) {
           currentOptions.splice(index, 1);
         }
-        this.setProperty(key, 'options', [...currentOptions]);
+        this.setProperty(key, 'options', [...currentOptions], otherForm);
```
Comment (trenthibbard) [inline projects/novo-elements/src/elements/form/FieldInteractionApi.ts:594]: Are these changes still necessary?

**Review by trenthibbard — CHANGES_REQUESTED** [review]

File: projects/novo-elements/src/elements/picker/extras/mixed-multi-picker-results/MixedMultiPickerResults.ts [hunk 6]
```diff
@@ -17,6 +17,9 @@ export interface IMixedMultiPickerOption {
         filterValue?: any,
     }[];
     getSecondaryOptionsAsync?(): Promise<{value: string, label: string}[]>;
+    // TODO: Refactor to prevent the need for a behaviorSubject to allow primaryOption's secondaryOptions to be cleared
+    // Currently secondaryOptions cannot be cleared via FieldInteraction API and must use a behavior subject - this includes modifyPickerConfig
+    clearSecondaryOptions?: BehaviorSubject<Boolean>;
```
[the remark this answers is not in the document]
Comment (EthanMcM) [reply projects/novo-elements/src/elements/picker/extras/mixed-multi-picker-results/MixedMultiPickerResults.ts]: Yes it can, added

File: projects/novo-elements/src/elements/form/FieldInteractionApi.ts (hunk 5 above)
Comment (EthanMcM) [reply projects/novo-elements/src/elements/form/FieldInteractionApi.ts:594]: These were added with a rebase

File: projects/novo-elements/src/elements/picker/extras/mixed-multi-picker-results/MixedMultiPickerResults.ts [hunk 7]
```diff
@@ -212,6 +228,11 @@ export class MixedMultiPickerResults extends BasePickerResults implements OnDest
                         this.inputElement.nativeElement.focus();
                     });
                 });
+                if (primaryOption.clearSecondaryOptions) {
+                    primaryOption.clearSecondaryOptions.subscribe(() => {
```
[the remark this answers is not in the document]
Comment (EthanMcM) [reply projects/novo-elements/src/elements/picker/extras/mixed-multi-picker-results/MixedMultiPickerResults.ts:235]: Will do, added in ngOnDestroy()

**Review by trenthibbard — APPROVED** [review]
